Bench with criterion

nemnem7 opened the pull request:
* Modified the body of functions which use criterion to do benchmark tests.
* Added sample size field to Cmd struct
* Renamed measurement_time field on Cmd struct to time
* Modified evm.rs file

rakita commented on bins/revme/src/cmd/bench/burntpix.rs:
```diff
@@ -28,7 +29,7 @@ sol! {
 }
 
-pub fn run() {
+pub fn run(criterion_group: &mut BenchmarkGroup<'_, WallTime>) {
```
Replacing `BenchmarkGroup` with `Criterion` seems better. There is no need to create a group at all

nemnem7 replied:
Done

rakita commented on bins/revme/src/cmd/bench/burntpix.rs:
```diff
@@ -45,33 +46,35 @@ pub fn run() {
-    svg(file_name, trimmed_data).expect("Failed to store svg");
+    criterion_group.bench_function("burntpix", |b| {
+        b.iter(|| {
+            let tx_result = evm.replay().unwrap().result;
+            let return_data = match tx_result {
+                ExecutionResult::Success {
+                    output, gas_used, ..
+                } => {
+                    println!("Gas used: {:?}", gas_used);
+                    match output {
+                        Output::Call(value) => value,
+                        _ => unreachable!("Unexpected output type"),
+                    }
+                }
+                _ => unreachable!("Execution failed: {:?}", tx_result),
+            };
+
+            // Remove returndata offset and length from output
+            let returndata_offset = 64;
+            let data = &return_data[returndata_offset..];
+
+            // Remove trailing zeros
+            let trimmed_data = data
+                .split_at(data.len() - data.iter().rev().filter(|&x| *x == 0).count())
+                .0;
+            let file_name = format!("{}_{}", seed, iterations);
+
+            svg(file_name, trimmed_data).expect("Failed to store svg");
```
Inside the `bench_function` we should use only `evm.replay()`.

Lets remove rest of the code after criterion and comment it out

nemnem7 replied:
Done

rakita commented on bins/revme/src/cmd/bench/snailtracer.rs:
```diff
@@ -19,13 +20,13 @@ pub fn simple_example(bytecode: Bytecode) {
+    criterion_group.bench_function("snailtracer", |b| {
+        b.iter(|| {
+            let bytecode = Bytecode::new_raw(Bytes::from(hex::decode(BYTES).unwrap()));
+            simple_example(bytecode);
```
Lets remove simple_example and leave only `evm.replay.unwrap()` inside `bench_function`

nemnem7 replied:
Done

rakita commented on bins/revme/src/cmd/bench/transfer.rs:
```diff
@@ -1,29 +1,24 @@
+pub fn run(criterion_group: &mut BenchmarkGroup<'_, WallTime>) {
+    criterion_group.bench_function("transfer", |b| {
+        b.iter(|| {
+            let mut evm = Context::mainnet()
+                .with_db(BenchmarkDB::new_bytecode(Bytecode::new()))
+                .modify_tx_chained(|tx| {
+                    // Execution globals block hash/gas_limit/coinbase/timestamp..
+                    tx.caller = BENCH_CALLER;
+                    tx.kind = TxKind::Call(BENCH_TARGET);
+                    tx.value = U256::from(10);
+                })
+                .build_mainnet();
```
Lets leave `evm.replay()` only in `bench_function` and move rest of code before

nemnem7 replied:
Done

rakita commented on bins/revme/src/cmd/bench.rs:
```diff
@@ -36,16 +36,45 @@ impl BenchName {
-            BenchName::Snailtracer => snailtracer::run(),
-            BenchName::Transfer => transfer::run(),
+            BenchName::Analysis => {
+                let mut criterion_group = criterion.benchmark_group("revme");
```
There is no need to create a benchmark group and we can send criterion to the ::run functions

nemnem7 replied:
Done

nemnem7 commented on bins/revme/src/cmd/bench.rs:
```diff
@@ -36,16 +36,45 @@ impl BenchName {
+    pub warmup: Option<f64>,
+    #[arg(short = 'm', long)]
+    pub time: Option<f64>,
+    #[arg(short = 's', long)]
+    pub samples: usize,
```
Done

rakita commented on bins/revme/src/cmd/bench.rs:
```diff
@@ -36,16 +36,40 @@ impl BenchName {
+            .warm_up_time(std::time::Duration::from_secs_f64(
+                self.warmup.unwrap_or(0.1),
+            ))
+            .measurement_time(std::time::Duration::from_secs_f64(self.time.unwrap_or(0.1)))
```
Would be good to add comment above on why 0.1s and, how sample/time interact together.

nemnem7 replied:
Done

rakita commented on bins/revme/src/cmd/bench.rs:
```diff
@@ -36,16 +36,40 @@ impl BenchName {
     pub fn run(&self) {
+        let mut criterion = criterion::Criterion::default()
+            .warm_up_time(std::time::Duration::from_secs_f64(
+                self.warmup.unwrap_or(0.1),
```
Should the default value be a little higher? Not sure but would say `0.5s` would be fine

nemnem7 replied:
Done

rakita commented on bins/revme/src/cmd/bench/burntpix.rs:
```diff
@@ -9,17 +10,19 @@ use static_data::{
 use database::{CacheDB, BENCH_CALLER};
 use revm::{
-    context_interface::result::{ExecutionResult, Output},
+    // context_interface::result::{ExecutionResult, Output},
```
```suggestion
```

nemnem7 replied:
Done.

rakita reviewed (approved):
few last nits but overall lgtm

nemnem7 commented:
@rakita

nemnem7 commented:
* Removed the imports which were previously commented out.
* Added comment which explains relationship and interaction between measurement_time and sample_size fields of Cmd struct
* Adjusted default warmup time to 0.5s (was previously 0.1s)
